jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java

```diff
@@ -150,90 +218,167 @@ protected Connection acquire(boolean create)
      * @param maxPending the max desired number of connections scheduled for opening,
      * or a negative number to always trigger the opening of a new connection
      */
-    protected void tryCreate(int maxPending)
+    Future<Void> tryCreate(int maxPending)
     {
-        while (true)
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        int pending = getPendingConnectionCount();
+        int total = pool.size();
+
+        if (LOG.isDebugEnabled())
+            LOG.debug("tryCreate {}/{} connections {}/{} pending", total, maxConnections, pending, maxPending);
+
+        if (total >= maxConnections)
         {
-            long encoded = connections.get();
-            int pending = AtomicBiInteger.getHi(encoded);
-            int total = AtomicBiInteger.getLo(encoded);
+            future.complete(null);
+            return future;
+        }
 
-            if (LOG.isDebugEnabled())
-                LOG.debug("tryCreate {}/{} connections {}/{} pending", total, maxConnections, pending, maxPending);
+        if (maxPending >= 0 && pending >= maxPending)
+        {
+            future.complete(null);
+            return future;
+        }
 
-            if (total >= maxConnections)
-                return;
+        if (LOG.isDebugEnabled())
+            LOG.debug("newConnection {}/{} connections {}/{} pending", total + 1, maxConnections, pending + 1, maxPending);
 
-            if (maxPending >= 0 && pending >= maxPending)
-                return;
+        Pool<Connection>.Entry entry = pool.addReservation();
```

Comment:
I would make the pool enforce max size, either preconfigured or passed in to `addReservation(maxConnections)`.  If it returns null here then you are over capacity so you can return the completed future.

Reply:
This both sounds very reasonable and natural for the pool to take care of its size. Let me do it.